tests/device_tests/test_pin.py

```diff
@@ -0,0 +1,118 @@
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the License along with this library.
+# If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.
+
+import time
+
+import pytest
+
+from trezorlib import messages
+from trezorlib.exceptions import PinException
+
+from ..common import get_test_address
+
+PIN4 = "1234"
+BAD_PIN = "5678"
+
+pytestmark = pytest.mark.setup_client(pin=PIN4)
+
+
+@pytest.mark.setup_client(pin=None)
+def test_no_protection(client):
+    with client:
+        client.set_expected_responses([messages.Address])
+        get_test_address(client)
+
+
+def test_correct_pin(client):
+    with client:
+        client.use_pin_sequence([PIN4])
+        # Expected responses differ between T1 and TT
+        is_t1 = client.features.model == "1"
+        client.set_expected_responses(
+            [
+                (is_t1, messages.PinMatrixRequest),
+                (
+                    not is_t1,
+                    messages.ButtonRequest(code=messages.ButtonRequestType.PinEntry),
+                ),
+                messages.Address,
+            ]
+        )
+        # client.set_expected_responses([messages.ButtonRequest, messages.Address])
+        get_test_address(client)
+
+
+@pytest.mark.skip_t2
+def test_incorrect_pin_t1(client):
+    with pytest.raises(PinException):
+        client.use_pin_sequence([BAD_PIN])
+        get_test_address(client)
+
+
+@pytest.mark.skip_t1
+def test_incorrect_pin_t2(client):
+    with client:
+        # After first incorrect attempt, TT will not raise an error, but instead ask for another attempt
+        client.use_pin_sequence([BAD_PIN, PIN4])
+        client.set_expected_responses(
+            [
+                messages.ButtonRequest(code=messages.ButtonRequestType.PinEntry),
+                messages.ButtonRequest(code=messages.ButtonRequestType.PinEntry),
+                messages.Address,
+            ]
+        )
+        get_test_address(client)
+
+
+def _check_backoff_time(attempts: int, start: float) -> None:
+    """Helper to assert the exponentially growing delay after incorrect PIN attempts"""
+    expected = (2 ** attempts) - 1
+    got = round(time.time() - start, 2)
+    assert got >= expected
+
+
+@pytest.mark.skip_t2
+def test_exponential_backoff_t1(client):
+    for attempt in range(3):
+        start = time.time()
+        with client, pytest.raises(PinException):
+            client.use_pin_sequence([BAD_PIN])
+            get_test_address(client)
+        _check_backoff_time(attempt, start)
+
+
+@pytest.mark.skip_t1
+def test_exponential_backoff_t2(client):
+    # Defining how long we expect each step to last from the beginning
+    # TODO: we should have some automatic way, but there seems to be no clear pattern
```

Comment:
This is because you are always checking against a fixed `START`, so the time consumed is cumulative.

The pattern is the same as with T1, so `device_wait_time = 2 ** attempt`, and `emulator_wait_time = device_wait_time - calculation_time`
The calculation time for T1 is 0 and for TT it seems to be 0.3.

Please modify so that the `start` time is updated before every attempt.
(you probably don't need to export the `start` variable out of the input_flow, the final `assert` can be left out.)

Reply:
Changed in 908b52d